src/sage/rings/semirings/tropical_variety.py

```diff
@@ -983,29 +1243,283 @@ def vertices(self):
         """
         if len(self._hypersurface) < 3:
             return set()
+        return set(self._vertices_components().keys())
 
-        vertices = set()
-        for i, component in enumerate(self._hypersurface):
-            parametric_function = component[0]
-            var = component[1][0].variables()[0]
-            interval = self._parameter_intervals()[i]
-            lower = interval[0].lower()
-            upper = interval[0].upper()
-            if lower != -infinity:
-                x = parametric_function[0].subs(**{str(var): lower})
-                y = parametric_function[1].subs(**{str(var): lower})
-                vertices.add((x, y))
-            if upper != infinity:
-                x = parametric_function[0].subs(**{str(var): upper})
-                y = parametric_function[1].subs(**{str(var): upper})
-                vertices.add((x, y))
-        return vertices
+    def _vertices_components(self):
+        """
+        Return the index of components adjacent to each vertex of ``self``.
+
+        OUTPUT:
+
+        A dictionary where the keys represent the vertices, and the
+        values are lists of tuples. Each tuple consists of the index
+        of an adjacent edge (component) `e_i` and a string indicating
+        the directionality of `e_i` relative to the vertex. The string
+        is either "pos" or "neg", specifying whether it is positive or
+        negative.
```

Comment:
Sorry for not realizing this earlier. It probably would be better to return `+1` / `-1` instead of `”pos”` / `”neg”` as it is easier to check.